File: day16/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var grid [][]byte

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, []byte(scanner.Text()))
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	height := len(grid)
	width := len(grid[0])

	fmt.Println("Part 1:", simulate(Beam{
		x:   0,
		y:   0,
		dir: DirEast,
	}, width, height, grid))

	fmt.Println("Part 2:", findLargest(width, height, grid))
}

func findLargest(width int, height int, grid [][]byte) int {
	greatest := 0
	for x := 0; x < width; x++ {
		m := simulate(Beam{
			x:   x,
			y:   0,
			dir: DirSouth,
		}, width, height, grid)
		greatest = max(greatest, m)
		m = simulate(Beam{
			x:   x,
			y:   height - 1,
			dir: DirNorth,
		}, width, height, grid)
		greatest = max(greatest, m)
	}
	for y := 0; y < height; y++ {
		m := simulate(Beam{
			x:   0,
			y:   y,
			dir: DirEast,
		}, width, height, grid)
		greatest = max(greatest, m)
		m = simulate(Beam{
			x:   width - 1,
			y:   y,
			dir: DirWest,
		}, width, height, grid)
		greatest = max(greatest, m)
	}
	return greatest
}

type (
	Dir int

	Beam struct {
		x, y int
		dir  Dir
	}

	Beams struct {
		beams []Beam
	}
)

func (b *Beams) Push(v Beam) {
	b.beams = append(b.beams, v)
}

func (b *Beams) Pop() (Beam, bool) {
	if len(b.beams) == 0 {
		return Beam{}, false
	}
	l := len(b.beams) - 1
	v := b.beams[l]
	b.beams = b.beams[:l]
	return v, true
}

const (
	DirNorth Dir = 0
	DirEast  Dir = 1
	DirSouth Dir = 2
	DirWest  Dir = 3
)

func simulate(start Beam, w, h int, grid [][]byte) int {
	hist := make([]bool, w*h)
	beamHist := make([]bool, w*h*4)

	sum := 0
	beams := &Beams{}
	beams.Push(start)
	for len(beams.beams) > 0 {
		beam, _ := beams.Pop()
		sum += stepBeam(w, h, beam, beams, grid, hist, beamHist)
	}

	return sum
}

func stepBeam(w, h int, beam Beam, beams *Beams, grid [][]byte, hist, beamHist []bool) int {
	hkey := beam.y*w + beam.x
	key := hkey*4 + int(beam.dir)
	if beamHist[key] {
		return 0
	}
	switch grid[beam.y][beam.x] {
	case '/':
		{
			next := beam
			switch next.dir {
			case DirNorth:
				next.x++
				next.dir = DirEast
			case DirEast:
				next.y--
				next.dir = DirNorth
			case DirSouth:
				next.x--
				next.dir = DirWest
			case DirWest:
				next.y++
				next.dir = DirSouth
			default:
				log.Fatalln("Invalid dir")
			}
			if isInBounds(next, w, h) {
				beams.Push(next)
			}
		}
	case '\\':
		{
			next := beam
			switch next.dir {
			case DirNorth:
				next.x--
				next.dir = DirWest
			case DirEast:
				next.y++
				next.dir = DirSouth
			case DirSouth:
				next.x++
				next.dir = DirEast
			case DirWest:
				next.y--
				next.dir = DirNorth
			default:
				log.Fatalln("Invalid dir")
			}
			if isInBounds(next, w, h) {
				beams.Push(next)
			}
		}
	case '|':
		{
			switch beam.dir {
			case DirNorth:
				{
					next := beam
					next.y--
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			case DirEast:
				{
					next := Beam{
						x:   beam.x,
						y:   beam.y - 1,
						dir: DirNorth,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
					next = Beam{
						x:   beam.x,
						y:   beam.y + 1,
						dir: DirSouth,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			case DirSouth:
				{
					next := beam
					next.y++
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			case DirWest:
				{
					next := Beam{
						x:   beam.x,
						y:   beam.y - 1,
						dir: DirNorth,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
					next = Beam{
						x:   beam.x,
						y:   beam.y + 1,
						dir: DirSouth,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			default:
				log.Fatalln("Invalid dir")
			}
		}
	case '-':
		{
			switch beam.dir {
			case DirNorth:
				{
					next := Beam{
						x:   beam.x - 1,
						y:   beam.y,
						dir: DirWest,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
					next = Beam{
						x:   beam.x + 1,
						y:   beam.y,
						dir: DirEast,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			case DirEast:
				{
					next := beam
					next.x++
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			case DirSouth:
				{
					next := Beam{
						x:   beam.x - 1,
						y:   beam.y,
						dir: DirWest,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
					next = Beam{
						x:   beam.x + 1,
						y:   beam.y,
						dir: DirEast,
					}
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			case DirWest:
				{
					next := beam
					next.x--
					if isInBounds(next, w, h) {
						beams.Push(next)
					}
				}
			default:
				log.Fatalln("Invalid dir")
			}
		}
	default:
		{
			next := beam
			switch next.dir {
			case DirNorth:
				next.y--
			case DirEast:
				next.x++
			case DirSouth:
				next.y++
			case DirWest:
				next.x--
			default:
				log.Fatalln("Invalid dir")
			}
			if isInBounds(next, w, h) {
				beams.Push(next)
			}
		}
	}
	beamHist[key] = true
	if hist[hkey] {
		return 0
	}
	hist[hkey] = true
	return 1
}

func isInBounds(beam Beam, w, h int) bool {
	return beam.x >= 0 && beam.y >= 0 && beam.x < w && beam.y < h
}

File: day16/src/main.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let grid = reader
        .lines()
        .map(|v| Ok::<_, Box<dyn std::error::Error>>(v?.into_bytes()))
        .collect::<Result<Vec<_>, _>>()?;

    let h = grid.len();
    let w = grid[0].len();

    println!(
        "Part 1: {}\nPart 2: {}",
        simulate(
            Beam {
                x: 0,
                y: 0,
                dir: Dir::East
            },
            w,
            h,
            &grid
        ),
        find_largest(w, h, &grid)
    );

    Ok(())
}

fn find_largest(w: usize, h: usize, grid: &Vec<Vec<u8>>) -> usize {
    (0..h).fold(
        (0..w).fold(0, |acc, x| {
            acc.max(simulate(
                Beam {
                    x,
                    y: 0,
                    dir: Dir::South,
                },
                w,
                h,
                grid,
            ))
            .max(simulate(
                Beam {
                    x,
                    y: h - 1,
                    dir: Dir::North,
                },
                w,
                h,
                grid,
            ))
        }),
        |acc, y| {
            acc.max(simulate(
                Beam {
                    x: 0,
                    y,
                    dir: Dir::East,
                },
                w,
                h,
                grid,
            ))
            .max(simulate(
                Beam {
                    x: w - 1,
                    y,
                    dir: Dir::West,
                },
                w,
                h,
                grid,
            ))
        },
    )
}

#[derive(Clone, Copy)]
enum Dir {
    North = 0,
    East = 1,
    South = 2,
    West = 3,
}

struct Beam {
    x: usize,
    y: usize,
    dir: Dir,
}

fn simulate(beam: Beam, w: usize, h: usize, grid: &Vec<Vec<u8>>) -> usize {
    let mut hist = vec![false; w * h];
    let mut beam_hist = vec![false; w * h * 4];

    let mut sum = 0;

    let mut beams = Vec::new();
    beams.push(beam);
    while let Some(beam) = beams.pop() {
        sum += step_beam(w, h, beam, &mut beams, grid, &mut hist, &mut beam_hist);
    }

    sum
}

fn step_beam(
    w: usize,
    h: usize,
    beam: Beam,
    beams: &mut Vec<Beam>,
    grid: &Vec<Vec<u8>>,
    hist: &mut Vec<bool>,
    beam_hist: &mut Vec<bool>,
) -> usize {
    let hkey = beam.y * w + beam.x;
    let key = hkey * 4 + beam.dir as usize;
    if beam_hist[key] {
        return 0;
    }
    match grid[beam.y][beam.x] {
        b'/' => {
            let mut next = beam;
            match next.dir {
                Dir::North => {
                    next.x += 1;
                    next.dir = Dir::East;
                }
                Dir::East => {
                    next.y = next.y.wrapping_sub(1);
                    next.dir = Dir::North;
                }
                Dir::South => {
                    next.x = next.x.wrapping_sub(1);
                    next.dir = Dir::West;
                }
                Dir::West => {
                    next.y += 1;
                    next.dir = Dir::South;
                }
            }
            if is_in_bounds(&next, w, h) {
                beams.push(next);
            }
        }
        b'\\' => {
            let mut next = beam;
            match next.dir {
                Dir::North => {
                    next.x = next.x.wrapping_sub(1);
                    next.dir = Dir::West;
                }
                Dir::East => {
                    next.y += 1;
                    next.dir = Dir::South;
                }
                Dir::South => {
                    next.x += 1;
                    next.dir = Dir::East;
                }
                Dir::West => {
                    next.y = next.y.wrapping_sub(1);
                    next.dir = Dir::North;
                }
            }
            if is_in_bounds(&next, w, h) {
                beams.push(next);
            }
        }
        b'|' => match beam.dir {
            Dir::North => {
                let mut next = beam;
                next.y = next.y.wrapping_sub(1);
                if is_in_bounds(&next, w, h) {
                    beams.push(next);
                }
            }
            Dir::East | Dir::West => {
                {
                    let next = Beam {
                        x: beam.x,
                        y: beam.y.wrapping_sub(1),
                        dir: Dir::North,
                    };
                    if is_in_bounds(&next, w, h) {
                        beams.push(next);
                    }
                }
                {
                    let next = Beam {
                        x: beam.x,
                        y: beam.y + 1,
                        dir: Dir::South,
                    };
                    if is_in_bounds(&next, w, h) {
                        beams.push(next);
                    }
                }
            }
            Dir::South => {
                let mut next = beam;
                next.y += 1;
                if is_in_bounds(&next, w, h) {
                    beams.push(next);
                }
            }
        },
        b'-' => match beam.dir {
            Dir::North | Dir::South => {
                {
                    let next = Beam {
                        x: beam.x.wrapping_sub(1),
                        y: beam.y,
                        dir: Dir::West,
                    };
                    if is_in_bounds(&next, w, h) {
                        beams.push(next);
                    }
                }
                {
                    let next = Beam {
                        x: beam.x + 1,
                        y: beam.y,
                        dir: Dir::East,
                    };
                    if is_in_bounds(&next, w, h) {
                        beams.push(next);
                    }
                }
            }
            Dir::East => {
                let mut next = beam;
                next.x += 1;
                if is_in_bounds(&next, w, h) {
                    beams.push(next);
                }
            }
            Dir::West => {
                let mut next = beam;
                next.x = next.x.wrapping_sub(1);
                if is_in_bounds(&next, w, h) {
                    beams.push(next);
                }
            }
        },
        _ => {
            let mut next = beam;
            match next.dir {
                Dir::North => {
                    next.y = next.y.wrapping_sub(1);
                }
                Dir::East => {
                    next.x += 1;
                }
                Dir::South => {
                    next.y += 1;
                }
                Dir::West => {
                    next.x = next.x.wrapping_sub(1);
                }
            }
            if is_in_bounds(&next, w, h) {
                beams.push(next);
            }
        }
    }
    beam_hist[key] = true;
    if hist[hkey] {
        return 0;
    }
    hist[hkey] = true;
    return 1;
}

fn is_in_bounds(beam: &Beam, w: usize, h: usize) -> bool {
    beam.x < w && beam.y < h
}
